std/string.d

```diff
@@ -2033,6 +2040,216 @@ S[] splitLines(S)(S s, in KeepTerminator keepTerm = KeepTerminator.no) @safe pur
+                                    goto default;
+                        }
+                        else
+                        {
+                            case lineSep:
+                            case paraSep:
+                                goto case '\n';
+                        }
+
+                        default:
+                            break;
+                    }
+                }
+            }
+            return _input[iStart .. iEnd];
+        }
+
+        void popFront()
+        {
+            if (iStart == _unComputed)
+            {
+                assert(!empty);
+                front();
+            }
+            iStart = _unComputed;
+        }
+
+        static if (isForwardRange!Range)
+        {
+            @property typeof(this) save()
+            {
+                auto ret = this;
+                ret._input = _input.save;
+                return ret;
+            }
+        }
+    }
+
+    return Result(r);
+}
+
+@safe pure unittest
+{
+    import std.conv : to;
+    import std.array : array;
+
+    debug(string) trustedPrintf("string.splitLines.unittest\n");
+
+    import std.exception;
+    assertCTFEable!(
+    {
+    foreach (S; TypeTuple!(char[], wchar[], dchar[], string, wstring, dstring))
+    {
+        auto s = to!S("\rpeter\n\rpaul\r\njerry\u2028ice\u2029cream\n\nsunday\nmon\u2030day\n");
+
+        auto lines = s.splitterLines().array;
+        assert(lines.length == 10);
+        assert(lines[0] == "");
+        assert(lines[1] == "peter");
+        assert(lines[2] == "");
+        assert(lines[3] == "paul");
+        assert(lines[4] == "jerry");
+        assert(lines[5] == "ice");
+        assert(lines[6] == "cream");
+        assert(lines[7] == "");
+        assert(lines[8] == "sunday");
+        assert(lines[9] == "mon\u2030day");
+
+        ubyte[] u = ['a', 0xFF, 0x12, 'b'];     // invalid UTF
+        auto ulines = splitterLines(cast(char[])u).array;
+        assert(cast(ubyte[])(ulines[0]) == u);
+
+        lines = splitterLines!(KeepTerminator.yes)(s).array;
+        assert(lines.length == 10);
+        assert(lines[0] == "\r");
+        assert(lines[1] == "peter\n");
+        assert(lines[2] == "\r");
+        assert(lines[3] == "paul\r\n");
+        assert(lines[4] == "jerry\u2028");
+        assert(lines[5] == "ice\u2029");
+        assert(lines[6] == "cream\n");
+        assert(lines[7] == "\n");
+        assert(lines[8] == "sunday\n");
+        assert(lines[9] == "mon\u2030day\n");
+
+        s.popBack(); // Lop-off trailing \n
+        lines = splitterLines(s).array;
+        assert(lines.length == 10);
+        assert(lines[9] == "mon\u2030day");
+
+        lines = splitterLines!(KeepTerminator.yes)(s).array;
+        assert(lines.length == 10);
+        assert(lines[9] == "mon\u2030day");
+    }
+    });
+}
+
+@nogc @safe pure unittest
+{
+    auto s = "\rpeter\n\rpaul\r\njerry\u2028ice\u2029cream\n\nsunday\nmon\u2030day\n";
+    auto lines = s.splitterLines();
+}
```

Comment:
An iteration would be nice here - after that we can pull this.

```d
static immutable witness = ["peter", "paul", "jerry", "\u2028ice\u2029cream", "sunday", "mon\u2030day"];
uint i;
foreach (line; lines)
{
    assert(line == witness[i++]);
} 
assert(i == 6);
```

After that we can pull this.